### ISSUE TITLE
`.groupby(..., maintain_order=True)` doesn't always maintain order.

### ISSUE THREAD

**Issue by coady** [pola-rs/polars#2112]
#### Are you using Python or Rust?
**Python**.

#### What version of polars are you using?
0.12.2

#### What operating system are you using polars on?
macOS 11.6

#### Describe your bug.
The `DataFrame.groupby` method has a `maintain_order` option, but it's not maintaining order in some cases. The pattern seems to be that it works with the `agg` method and expressions, but not with the higher-level `GroupBy` methods. 

#### What are the steps to reproduce the behavior? What is the actual behavior?
```python
In []: df = pl.DataFrame({'x': list('bab'), 'y': range(3)})

In []: df.groupby('x', maintain_order=True).agg(pl.count('y'))
Out[]: 
shape: (2, 2)
┌─────┬─────────┐
│ x   ┆ y_count │
│ --- ┆ ---     │
│ str ┆ u32     │
╞═════╪═════════╡
│ b   ┆ 2       │
├╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ a   ┆ 1       │
└─────┴─────────┘

In []: df.groupby('x', maintain_order=True).agg(pl.list('y'))
Out[]: 
shape: (2, 2)
┌─────┬────────────┐
│ x   ┆ y_agg_list │
│ --- ┆ ---        │
│ str ┆ list [i64] │
╞═════╪════════════╡
│ b   ┆ [0, 2]     │
├╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┤
│ a   ┆ [1]        │
└─────┴────────────┘
```
So far so good, but the same code with `count` and `agg_list` (intermittently) reorders the groups.
```python
In []: df.groupby('x', maintain_order=True).count()
Out[]: 
shape: (2, 2)
┌─────┬─────────┐
│ x   ┆ y_count │
│ --- ┆ ---     │
│ str ┆ u32     │
╞═════╪═════════╡
│ a   ┆ 1       │
├╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ b   ┆ 2       │
└─────┴─────────┘

In []: df.groupby('x', maintain_order=True).agg_list()
Out[]: 
shape: (2, 2)
┌─────┬────────────┐
│ x   ┆ y_agg_list │
│ --- ┆ ---        │
│ str ┆ list [i64] │
╞═════╪════════════╡
│ a   ┆ [1]        │
├╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┤
│ b   ┆ [0, 2]     │
└─────┴────────────┘
```

#### What is the expected behavior?
Always maintaining order of the group keys.

**Comment by ritchie46** [comment]
Oh.. right they still dispatch to an older implementation without expressions.

What they need to be is syntactic sugar for `agg(pl.all().aggregation()`.

E.g. 

```python
df.groupby().first()
```

should desugar to:

```python
(df.groupby()
   .agg(pl.all().first()
)
```

This can be implemented completely from python. I will tag this as `good_first_issue` for anybody willing to take this up.

**Comment by mhconradt** [comment]
I'm going to try to knock this one out this afternoon

**Comment by mhconradt** [comment]
While I'm at it, also going to take care of #1722

**Comment by mhconradt** [comment]
@ritchie46 I noticed `agg_list` isn't supported by the expressions API. Was this intentionally omitted?

**Comment by ritchie46** [comment]
Fixed in  #2136
